Modules/socketmodule.c

```diff
@@ -5470,6 +5470,61 @@ AF_UNIX if defined on the platform; otherwise, the default is AF_INET.");
 #endif /* HAVE_SOCKETPAIR */
 
 
+/* socket.fdtype() function */
+
+#if defined(SO_TYPE) && !defined(MS_WINDOWS)
+/* set if we can implement fdtype().  On Windows, getsockname() fails with
+   error 10022.  There may be other platforms that have SO_TYPE but also
+   don't provide the necessary functionality. In those cases the #if above
+   may need tweaking. */
+#define HAVE_FDTYPE
+#endif
+
+#ifdef HAVE_FDTYPE
+static PyObject *
+socket_fdtype(PyObject *self, PyObject *fdobj)
+{
+    SOCKET_T fd;
+    int sock_type;
+    struct sockaddr sa;
+    socklen_t l;
+    int protocol;
+
+    fd = PyLong_AsSocket_t(fdobj);
+    if (fd == (SOCKET_T)(-1) && PyErr_Occurred())
+        return NULL;
+
+    l = sizeof(sock_type);
+    if (getsockopt(fd, SOL_SOCKET, SO_TYPE, &sock_type, &l) < 0) {
+        return set_error();
+    }
+
+    l = sizeof(sa);
+    if (getsockname(fd, &sa, &l) < 0) {
```

Comment:
Your code doesn't work on Windows for IPv6 addresses. You have to make the buffer large enough to fit all possible address types in it.

```
sock_addr_t addrbuf;
socklen_t addrlen = sizeof(sock_addr_t);

memset(&addrbuf, 0, addrlen);
if (getsockname(fd, SAS2SA(&addrbuf), &addrlen) == 0) {
    family = SAS2SA(&addrbuf)->sa_family;
} else {
    return set_error();
}
```